Implement Realm Database

rajivshah3 opened the pull request:
<!---
Hints for a successful PR:
1. It is recommended that before you submit a PR to IOTA, to open an issue first and assign yourself.
This way you may get inputs and discover parallel PRs to the one you want to submit.
2. In case of a big PR, consider breaking it up into smaller PRs. This will help to get it merged in an incremental process.
3. Note that a PR should have a *single* area of responsibility. If your PR does more than one thing than it should be split into several PRs!!!!!
4. It will be helpful if you make additional comments on the code via GitHub PR review to explain the choices you made
-->

# Description

Closes #311 

## Type of change

- Enhancement (a non-breaking change which adds functionality)


# How Has This Been Tested?

- Test A
- Test B


# Checklist:

- [ ] My code follows the style guidelines for this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] For changes to `mobile` that include native code (including React Native modules): I have verified that both iOS and Android successfully build in both `Debug` and `Release` modes
- [ ] For changes to `shared`: If applicable, I have verified that my changes are implemented correctly in `desktop` and `mobile`

rajivshah3 committed:
Mobile: Add and link Realm

rajivshah3 committed:
Shared: Define some schemas

rajivshah3 committed:
Update schemas

rajivshah3 committed:
Merge branch 'develop' into feature/realm

rajivshah3 committed:
Mobile/iOS: Update pbxproj

rajivshah3 committed:
Merge branch 'develop' into feature/realm

rajivshah3 committed:
Move realm to shared

rajivshah3 committed:
Shared: Store chart data in realm

rajivshah3 committed:
Merge branch 'develop' into feature/realm

rajivshah3 committed:
Bug fix

Commit:
Change input selection

Commit:
Merge develop into feat/input-selection

Commit:
Add coverage for #prepareInputs

rajivshah3 committed:
Merge branch 'develop' into feature/realm

rajivshah3 committed:
Fix lockfile

Commit:
Clean up #getInputs and related utils

Commit:
Remove unused utils #getOutgoingTransfersForAddresses & #getPendingOutgoingTransfersForAddresses

Commit:
Update coverage for libs/iota/addresses

Commit:
Add coverage for #getInputs

Commit:
Fix failing tests in libs/iota/transfers

Commit:
Merge branch 'develop' into feat/input-selection

Commit:
Update test coverage for actions/transfers

rajivshah3 committed:
Merge branch 'develop' into feature/realm

Commit:
Remove realm & redux-persist-realm dependency from mobile directory

Commit:
Avoid passing Realm constructor to state actions

Commit:
Setup base for realm

rajivshah3 committed:
Merge pull request #490 from laumair/feat/realm

Realm setup

rajivshah3 committed:
Merge branch 'develop' into feature/realm

rajivshah3 committed:
Merge branch 'develop' into feat/input-selection

rajivshah3 committed:
Merge branch 'develop' into feature/realm

rajivshah3 committed:
Merge branch 'develop' into feature/realm

Commit:
Update schema

- Add Wallet schema
- Add Account schema
- Update Transaction schema
- Update Address schema

Commit:
Add a selector for selecting bundles for auto promotion

Commit:
Add some storage (realm) to state (redux) mappers

Commit:
Consume storage to state mapper for restoring state when mobile app intialises

Commit:
Allow purging realm storage on wallet reset

Commit:
Migrate to realm for account state for a new/fresh account

Commit:
Migrate account deletion & account name change to realm

Commit:
Merge develop into feat/realm-migration

Commit:
Integrate realm with mobile's entry point

Commit:
Integrate account syncs, polling, snapshot transition with realm storage

Commit:
Migrate accounts (transaction & addresses) related data management from AsyncStorage to realm

Commit:
Merge develop into feat/input-selection

Commit:
Merge remote tracking branch feat/input-selection

Commit:
Add maxInputs property to SeedVault class and use it during input selection

Commit:
Merge develop into feature/realm

rajivshah3 committed:
Mobile: Relink Realm

Commit:
Merge branch feat/input-selection into feat/realm-migration

Commit:
Add realm binaries to gitignore

Commit:
Clean up transaction utils and update test coverage

- Remove transaction util #categoriseTransactionsByPersistence
- Remove transaction util #transformTransactionsByBundleHash
- Remove transaction util #isValidTransactionSync
-  Remove transaction util #isValidTransactionAsync
- Remove transaction util #filterInvalidTransactionsSync
- Remove transaction util #filterInvalidTransactionsAsync
- Remove transaction util #prepareForAutoPromotion
- Remove transaction util #getPendingTxTailHashes
- Remove transaction util #markTransfersConfirmed
- Remove transaction util #getConfirmedTransactionHashes
- Remove transaction util #mergeNewTransfers
- Remove transaction util #getBundleHashesForNewlyConfirmedTransactions
- Remove transaction util #isStillAValidTransaction
- Remove transaction util #getOwnTransactionHashes
- Remove transaction util #pickNewTailTransactions

Commit:
Remove unnecessary tests for reducers/accounts

Commit:
Make maxInputs a read-only property of SeedVault class

Commit:
Fix some unit tests for libs/iota/addresses

Commit:
Remove unnecessary check for remainder transaction when inputs are categorised

Some bundles (e.g., MECOSAGPFIHBAJQBBX9HUJRNJEKRVNRJBHEBWYPD9H9IKBFEJZCRLLFYLLILEUHCUNYEHBEQARWNHFWBC)
are not properly categorised into inputs/outputs. This is because input transactions of such bundles have currentIndex === lastIndex.

This commit removes the unnecessary checks on inputs (restricting them to be non-remainder transactions). Categorisation of a bundle to inputs/outputs should only be based on negative/non-negative value.

 # Please enter the commit message for your changes. Lines starting

rajivshah3 committed:
[Realm] Node schema (#499)

* Shared: Add node schema

* Refactor and add tests

* Remove unnecessary import

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge branch 'develop' into feat/input-selection

Commit:
Make maxInputs a read-only property of SeedVault class

Commit:
Merge upstream branch feature/realm into feat/realm-migration

Commit:
Migrate settings redux reducer to realm

Commit:
Simplify theming state management

Currently, we store both theme object and theme name in settings reducer. This commit simplifies the theme setup by removing theme object from state and using a state selector for passing active theme object to all UI components. This also avoids storing theme object in persistent storage.

Commit:
Remove ChartDataSchema, DataForTimeframeSchema, DataPointSchema

Commit:
Use theme selector for injecting theme object to all UI components (mobile & shared)

Commit:
Migrate node and currency related state management to realm

Commit:
Update test coverage for settings reducer

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge branch develop into feat/input-selection

Commit:
Fix typo

Commit:
Merge pull request #480 from laumair/feat/input-selection

Update input selection process

Commit:
Fix test coverage for libs/iota/addresses

Commit:
Merge branch develop into branch next

Commit:
Fix transactions schema and related tests

Commit:
Merge pull request #516 from laumair/feat/realm-migration

Use realm as persistent storage

Commit:
Add auto migration from AsyncStorage to Realm storage

This commit adds the ability for the wallet to automatically migrate (accounts & settings) data from old (AsyncStorage) to Realm storage. After login, users will automatically be redirected to the migration screen (on mobile) where behind the scenes, auto migration will be performed. On successful migration, users will be redirected to the dashboard.

Commit:
Merge branch develop into branch next

Commit:
Merge pull request #600 from laumair/feat/realm-migration

Add auto migration from AsyncStorage to Realm storage

Commit:
Merge branch 'develop' into next

Commit:
Merge branch next into feature/realm

Commit:
Minor fixes

- Fix balance display for accounts in TopBar (mobile)
- Fix reference to address data in ViewAddresses (mobile)
- Fix some failing tests
- Refactor tests for actions/transfers/makeTransaction

Commit:
Minor fixes

- Fix failing tests
- Fix some iota utils

Commit:
Make sure getInput stubs return an object and not an array of inputs

rajivshah3 committed:
Upgrade Snyk and whitelist vulnerability (#630)

* Upgrade Snyk

* Whitelist vulnerability, this cannot be easily patched and Trinity is not vulnerable

rajivshah3 committed:
Upgrade to React Native 0.57 (#438)

* Mobile: Update to React Native 0.56

* Mobile: Clean up RN update

* Mobile: Remove react-native-keyboard-aware-scroll-view

* Mobile: Fix Android keyboard avoidance

* Mobile: Fix chart interpolation Android

* Mobile: Fix Android account name cut-off

* Mobile: Upgrade React Native to 0.57

* Bug fixes

* Bug fixes

* Mobile: Fix hidden back button

* Update lockfile

* Mobile: Re-add JVM args

* Mobile: Add babel-plugin-jest-hoist to fix Bugsnag mocks

* Shared: Fix babel-related mocha issues

* Remove unnecessary patches

* Update lockfile

* Mobile: Apply forwardRef fix

* Mobile: Fix failing Balance test

* Mobile: Revert unintended changes

* Mobile: Re-add react-native-vector-icons patch

* Remove unnecessary babel files/deps

* Fix module-resolver, ignore some YellowBox warnings

* Shared: Revert redux-persist upgrade

* Make sure greenkeeper does not update redux-persist

* Shared: Revert React update

* Revert "Merge branch 'hotfix/revert-upgrade' into chore/rn-0.57"

This reverts commit 3b5a8e76934a5f20fdf911d8e38bc4ed9e33f22f, reversing
changes made to a6bbbbb95878f343e367ae7a78f5984edb584555.

* Mobile: Bump to React Native 0.57.5, React 16.6.1

* Revert unnecessary change [ci skip]

* Upgrade react-native-vector-icons

Closes #603

* Upgrade react-native-navigation

* Use new Xcode build system

* Mobile: Fix Android build

* Mobile: Fix failing test

* Mobile: Revert RNN upgrade, revert to legacy build system

* Mobile: Cherry pick wix/react-native-navigation@ab2f335 for RN 0.57 support

* Mobile: Fix typo in gradlew.bat

Commit:
Finalise Realm instantiation  (#777)

* Import realm from mobile directory instead of directly importing it from shared

* Finalize realm usage for desktop, mobile & test environments

* Disable realm analytics reports

Commit:
Merge branch develop into branch feature/realm

Commit:
Fix failing tests and warnings for mobile

Commit:
Use detectOpenHandles argument for jest tests

Commit:
Force exit mocha tests

Commit:
Fix incorrect refs & add realm initialisation setup for desktop

Commit:
Merge branch develop into branch feature/realm

Commit:
Show desktop window only after persisted state has been restored

Commit:
Send payload only if provided

Commit:
Merge branch develop into branch feature/realm

Commit:
Merge remote-tracking branch 'upstream/develop' into test/desktop-init

Commit:
Merge branch develop into branch feature/realm

Commit:
Merge branch test/desktop-init into branch feature/realm

rajivshah3 commented:
Blocked by https://github.com/realm/realm-js/issues/2180. The CI will hang indefinitely when tests are run

rajivshah3 committed:
Shared: Rebuild realm after installing dependencies (#842)

Commit:
Merge branch 'mobile-alpha/realm' into feature/realm

Commit:
Merge remote-tracking branch 'upstream/feature/realm' into feature/realm

Commit:
Remove undefined i18next translate function from progressSteps

cvarley100 commented on src/desktop/src/index.js:
```diff
@@ -30,80 +33,78 @@ export const bugsnagClient = bugsnag({
+        const data = JSON.parse(payload);
+        const statePartial = {};
+        statePartial[data.key.replace('reduxPersist:', '')] = data.item;
+        // persistor.rehydrate(statePartial, { serial: true });
```
Should this comment be removed?

laumair replied:
We are no longer using `redux-persist` so this should be removed. However, @rihardsgravis should confirm if we need to directly access persisted data from `Realm` at this point.

cvarley100 commented on src/mobile/src/ui/components/Poll.js:
```diff
@@ -141,14 +145,15 @@ export class Poll extends Component {
-                const bundles = keys(unconfirmedBundleTails);
-                const bundleHash = bundles[0];
+                const bundleHashes = keys(unconfirmedBundleTails);
+                const bundleHashToPromote = bundleHashes[random(size(bundleHashes) - 1)];
```
In theory, this could keep promoting the same bundles while missing others. We should just promote in order of oldest to latest imo.

laumair replied:
You're right. This is how it should be promoted ideally (oldest to latest). Adding that would require to maintain state. For now, I've added a `TODO` and will submit a separate PR for this.

cvarley100 commented on src/mobile/src/ui/views/wallet/Send.js:
```diff
@@ -683,18 +682,8 @@ export class Send extends Component {
-            isKeyboardActive,
-        } = this.props;
-        const textColor = { color: body.color };
+        const { t, isSendingTransfer, address, amount, message, denomination, theme, isKeyboardActive } = this.props;
```
There seems to be some inconsistency with theme variable use. In another area we have `{ body, primary, secondary } = theme;`. Is there a reason for each?

laumair replied:
I don't understand this point. `{ body, primary, secondary } = theme` is just destructuring the object here. The changes related to theming in this PR are different. Instead of storing/persisting the whole theme object, it just stores the `themeName` in state and uses a selector to reference the theme object.

cvarley100 commented on src/shared/__tests__/__samples__/addresses.js:
```diff
@@ -0,0 +1,94 @@
+const latestAddressObject = {
```
Nice cleanup

cvarley100 commented on src/shared/actions/settings.js:
```diff
@@ -157,12 +175,19 @@ const addCustomNodeRequest = () => ({
  * @param {string} payload
  *
- * @returns {{type: {string}, payload: {string} }}
+ * @returns {{type: {string}, url: {string}, remotePow: {string} }}
```
It doesn't return `remotePow`

laumair replied:
Fixed in https://github.com/iotaledger/trinity-wallet/pull/375/commits/5996a50a886055ac38bfac06b40b3029e808d468

cvarley100 commented on src/shared/libs/iota/accounts.js:
```diff
@@ -1,279 +1,137 @@
-                map(thisStateCopy.transfers, (tx) => tx),
-                keyIndexes,
-            );
+            // if (notificationFn) {
```
Why is this commented out?

laumair replied:
Fixed in https://github.com/iotaledger/trinity-wallet/pull/375/commits/5996a50a886055ac38bfac06b40b3029e808d468

cvarley100 commented on src/shared/locales/en/translation.json:
```diff
@@ -789,6 +767,11 @@
     },
+    "migration": {
+        "dataMigration": "Data migration",
+        "dataMigrationExplanation": "Due to performance and storage limitations, Trinity now uses a new database for storing your wallet settings and accounts data. Please wait while Trinity automatically migrates your wallet settings and account information.",
```
```suggestion
        "dataMigrationExplanation": "Due to performance and storage limitations, Trinity is using a new database for storing settings and account data. Please wait while Trinity automatically migrates your existing data.",
```

laumair replied:
Fixed in https://github.com/iotaledger/trinity-wallet/pull/375/commits/5996a50a886055ac38bfac06b40b3029e808d468

cvarley100 commented on src/shared/locales/en/translation.json:
```diff
@@ -789,6 +767,11 @@
+    "migration": {
+        "dataMigration": "Data migration",
+        "dataMigrationExplanation": "Due to performance and storage limitations, Trinity now uses a new database for storing your wallet settings and accounts data. Please wait while Trinity automatically migrates your wallet settings and account information.",
+        "problemMigratingYourData": "There was a probelm migrating your data. Please try again"
```
```suggestion
        "problemMigratingYourData": "There was a problem migrating your data. Please try again"
```

laumair replied:
Fixed in https://github.com/iotaledger/trinity-wallet/pull/375/commits/5996a50a886055ac38bfac06b40b3029e808d468

Commit:
Update prepateTransferArray implementation to accept addressData as an array instead of an object

Commit:
Merge branch mobile-alpha/realm into branch feature/realm

Commit:
Minor fixes and updates

- Remove manual state rehydration from src/desktop/src/index.js
- Pass in theme object to UnitInfoModal
- Fix JSDocs for addCustomNodeSuccess action creator
- Fix notificationFn trigger in syncAccount
- Relocate mapNormalisedTransactions util

Commit:
Always find transaction hashes diff from transactions with own addresses

Commit:
Update UI for migration screen

Commit:
Assign index & meta to account object during migration

Commit:
Use buildNumber for detecting & triggering redux->realm migration

rajivshah3 committed:
Shared: Ignore chownr vulnerability in Snyk

rihardsgravis committed:
Desktop: Realm Database implementation Desktop fixes (#874)

* Desktop related fixes:
- Fix balance setting in `Balance` and `Sidebar` components
- Fix latest address retrieval
- Fix notification function
- Fix List component transaction list retrieval

* Remove `reverse` from account address list

* Mobile: Resolve lint errors

* Revert 5291972 and fix eslint config (see eslint/eslint#11231)

Commit:
Fix failing tests for libs/iota/accounts

cvarley100 commented on src/shared/storage/index.js:
```diff
@@ -0,0 +1,790 @@
+
+    /**
+     *
+     * @
```
```suggestion
     * @method addNodes
```

rajivshah3 replied:
Fixed in c156cad

cvarley100 commented on src/shared/actions/migrations.js:
```diff
@@ -0,0 +1,126 @@
+ *  *
+ * @returns {{type: {string}, payload: {bool} }}
+ */
+export const setMigrationStatus = (payload) => {
```
Maybe best to use `setRealmMigrationStatus` here and update the associated variable names. We use the term `migration` for a couple of things.

laumair replied:
Fixed in https://github.com/iotaledger/trinity-wallet/pull/375/commits/4198d97852e19380c8d557b68a58a6a1a8490375

Commit:
Fix checksum for latestAddressObject & add a separate constant for latestAddressBalance

Commit:
Fix attachAndFormatAddresses util implementation

cvarley100 committed:
Mobile: Fix migration step strings

cvarley100 committed:
Mobile: Update prop types

rajivshah3 committed:
Shared: Fix documentation for delete and addNodes

Commit:
Remove hash as a primary key from Transaction schema

Commit:
Add realm-object-server/ to gitignore

Commit:
Merge remote-tracking branch 'upstream/feature/realm' into feature/realm

rajivshah3 committed:
Bump realm to v2.21.1

Commit:
Update Transaction schema

- Add attachmentTimestamp
- Add attachmentTimestampLowerBound
- Add attachmentTimestampUpperBound
- Add obsoleteTag

Commit:
Minor fixes

- Map correct persistence to normalised transactions
- Fix parameters for constructBundleFromTransactions util

Commit:
Merge remote-tracking branch 'upstream/feature/realm' into feature/realm

Commit:
Minor fixes

- Make completedMigration a required prop in Login & Migration component
- Make sure missing properties like completedMigration are correctly mapped to redux store on entry

Commit:
Include version check on app entry

Commit:
Minor fixes and updates

- Update method description for realm Wallet class method updateLatest
- Rename setMigrationStatus action creator to setRealmMigrationStatus

rihardsgravis committed:
Desktop: Realm Database - remove redux persist settings dependency (#877)

* Update tray application state sync and initialisation

* Update Proxy settings to use separate electronSettings entry

rajivshah3 committed:
Shared: Fix periodically failing sortTransactionTrytesArray test (#878)

* Shared: Fix periodically failing sortTransactionTrytesArray test

* Shared: Make recommended changes

* Remove unnecessary assert statements

Commit:
Merge remote-tracking branch 'upstream/develop' into feature/realm

Commit:
Merge branch 'feature/realm' of github.com:iotaledger/trinity-wallet into feature/realm

Commit:
Merge branch 'develop' into feature/realm

Commit:
Merge branch 'develop' into feature/realm

Commit:
Fix skipped tests for #isNodeHealthy

Commit:
Fix skipped tests

Commit:
Update build number check realm migration detection

Commit:
Update build number to 40 for migration detection check

Commit:
Add missing getCustomNodesFromState state selector

Commit:
Pass nodes array in correct format to quorum methods

Commit:
Migrate accountIndex property for account from AsyncStorage to realm

Commit:
Remove primary key (address) from AddressSchema

Commit:
Fix invalid bundle construction for failed transactions

Commit:
Preserve local spend status before updating account data in realm

Commit:
Merge branch 'develop' into feature/realm

Commit:
Refactor #getFullAddressHistory tests

Commit:
Make sure we fetch persisted account indexes from realm

Commit:
Check for undefined addressData prop before updating address data in realm

Commit:
Merge branch 'develop' into feature/realm

Commit:
Realm data encryption (#1018)

* Add base setup for realm data encryption

* Pass #getEncryptionKeyPromise when storage is reinitialised

* Store realm encryption key in keychain

* Initialise realm instance with encryption key in tests

* Store realm encryption key in keychain

* Address comments

- Remove base64-js
- Perform Uint8Array to string conversions with vanilla JS

* Skip realm encryption key to be reset on password change

* Simplify #getEncryptionKey implementation

* Update build number to 41 for migration detection check

* Mobile: Fix notification icon touch radius

* Include isRetryingFailedTransaction in modalProps when modalProps are updated

* Move manual bundle construction implementation (for failed transactions) in constructBundlesFromTransactions

* Mobile: Fix iPhone X modal visual bug

Commit:
Merge branch develop into branch feature/realm

Commit:
Merge remote-tracking branch 'upstream/feature/realm' into feature/realm

rajivshah3 committed:
Merge branch 'develop' into feature/realm

cvarley100 committed:
Mobile Release 0.4.1 (41) (#1029)

* Mobile: Bump build number to 41

* Mobile: Bump Realm migration versioning

rajivshah3 committed:
Merge branch 'develop' into feature/realm

rajivshah3 committed:
Shared: Only rebuild Realm on Debian

Commit:
Correctly assign new account name in realm storage (#1045)

Commit:
Update isFailedTransaction prop when modal props are updated (#1046)

cvarley100 committed:
Mobile: Add retry button, error log and change node to Realm migration (#1041)

* Mobile: Add ability to change node and retry during migration

* Mobile: Address comments

* Mobile: Fix notification button import and padding

* Mobile: Disable iOS pop gesture

cvarley100 committed:
Mobile Release 0.6.1 (42) (#1048)

* Mobile: Bump build no to 42

* Mobile: Update realm migration versioning

rihardsgravis committed:
Realm Database implementation desktop bugfixes (#1025)

* - Fix Realm storage path
- Add missing wallet reset triggers
- Fix address component prop use
- Remove failed bundle hash action

* Update Realm path for test environment

* - Remove Realm instance init from Tray application
- Keep Realm encryption key on keychain initialisation

* Focus wallet window after initial store update

* Add missing Windows required dependency

Commit:
Merge branch 'develop' into feature/realm

rajivshah3 committed:
Desktop: Create Migration component to migrate data (#857)

* Desktop: Create Migration component to migrate data

* Desktop: Add 'history' to PropTypes

* Desktop: Document getAllStorageKeys

* - Fix Realm storage path
- Add missing wallet reset triggers
- Fix address component prop use
- Remove failed bundle hash action

* Update Realm path for test environment

* - Remove Realm instance init from Tray application
- Keep Realm encryption key on keychain initialisation

* Focus wallet window after initial store update

* Add missing Windows required dependency

* Move Migration to `ui/global`

* Migration bugfixes

* Desktop: Fix off-by-one mistake on Migration component

* Code review fixes

* Fix Wallet reset functionallity

Commit:
Fix account duplication on account rename operation (#1077)

Related issue: https://github.com/iotaledger/trinity-wallet/issues/1066

cvarley100 committed:
Mobile: Readd react-native-translucent-moddal (#1083)

Commit:
Fix invalid address data issue (#1089)

Interrupting new account onboarding (on loading screen) leads wallet to throw continuous exceptions. The reason for exceptions was missing "completed" property in realm schema. This commit fixes the issue and also adds realm migration from schema version 0 to latest schema.

cvarley100 committed:
Resolve conflicts

rihardsgravis committed:
Add migration retry and node change functionality (#1094)

rihardsgravis committed:
Desktop: Realm implementation fixes (#1096)

* - Exclude unnecessary realm package contents in builds
- Fix unique seed check to exclude Realm key
- Fix account snapshot transition missing addresses

* Missing tag build error fix

rihardsgravis committed:
Desktop: Updated Entangled node use (#1095)

* Implement new Entangled node bindings

* Update shrinkwrap file

cvarley100 committed:
Mobile: Link translucent modal

cvarley100 committed:
Merge branch 'feature/realm' of https://github.com/iotaledger/trinity-wallet into feature/realm

Commit:
Merge branch 'develop' into feature/realm

* develop:
  Bump react from 16.8.1 to 16.8.2 in /src/desktop (#1075)
  Bump @ledgerhq/hw-transport-node-hid in /src/desktop (#1082)
  [Security] Bump braces from 1.8.5 to 2.3.2 in /src/desktop (#1079)
  Bump i18next from 15.0.2 to 15.0.4 in /src/desktop (#1072)
  Bump react-dom from 16.8.1 to 16.8.2 in /src/desktop (#1076)
  Desktop Ledger app update (#1059)

# Conflicts:
#	src/desktop/npm-shrinkwrap.json
#	src/desktop/package.json
#	src/shared/actions/transfers.js

Commit:
Fix desktop transition `Cannot read property 'type' of undefined` error

Commit:
Update shrinkwrap file

rajivshah3 committed:
Shared: Fix migration when app is first installed (#1098)

rihardsgravis committed:
- Fix receive closes automatically on Ledger account (#1101)

- Fix unable to cancel refreshing history without Ledger connected
- Fix onboarding seed not available after failed initial account fetch

cvarley100 committed:
Resolve conflicts
